# Russell Data CLeaning and EDA

## Imports and Helper functions

In [2]:
from warnings import simplefilter
simplefilter(action='ignore')

from IPython.display import Markdown, display
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn 
import pandas as pd
import numpy as np
import sympy as sympy
import time 
import enum

In [82]:
#Clean method to return all the image IDs of a given patient ID
def returnAllPatientImages(patient):
    return(list(train_data[train_data["PatientID"]==patient]["StudyInstanceUID"]))

## Data Cleanining and Integrity Checking

In [5]:
train_data = pd.read_csv("train - train.csv")
train_data

StudyInstanceUID  ETT - Abnormal  \
0      1.2.826.0.1.3680043.8.498.26697628953273228189...               0   
1      1.2.826.0.1.3680043.8.498.46302891597398758759...               0   
2      1.2.826.0.1.3680043.8.498.23819260719748494858...               0   
3      1.2.826.0.1.3680043.8.498.68286643202323212801...               0   
4      1.2.826.0.1.3680043.8.498.10050203009225938259...               0   
...                                                  ...             ...   
30078  1.2.826.0.1.3680043.8.498.74257566841157531124...               0   
30079  1.2.826.0.1.3680043.8.498.46510939987173529969...               0   
30080  1.2.826.0.1.3680043.8.498.43173270582850645437...               0   
30081  1.2.826.0.1.3680043.8.498.95092491950130838685...               0   
30082  1.2.826.0.1.3680043.8.498.99518162226171269731...               0   

       ETT - Borderline  ETT - Normal  NGT - Abnormal  NGT - Borderline  \
0                     0             0               0                 0   
1                     0             1               0                 0   
2                     0             0               0                 0   
3                     0             0               0                 0   
4                     0             0               0                 0   
...                 ...           ...             ...               ...   
30078                 0             1               0                 0   
30079                 0             0               0                 0   
30080                 0             1               0                 0   
30081                 0             0               0                 0   
30082                 0             1               0                 0   

       NGT - Incompletely Imaged  NGT - Normal  CVC - Abnormal  \
0                              0             1               0   
1                              1             0               0   
2                              0             0               0   
3                              0             0               1   
4                              0             0               0   
...                          ...           ...             ...   
30078                          0             0               0   
30079                          0             0               0   
30080                          1             0               1   
30081                          0             0               0   
30082                          0             0               0   

       CVC - Borderline  CVC - Normal  Swan Ganz Catheter Present  PatientID  
0                     0             0                           0  ec89415d1  
1                     0             1                           0  bf4c6da3c  
2                     1             0                           0  3fc1c97e5  
3                     0             0                           0  c31019814  
4                     0             1                           0  207685cd1  
...                 ...           ...                         ...        ...  
30078                 1             1                           0  5b5b9ac30  
30079                 0             1                           0  7192404d8  
30080                 0             1                           0  d4d1b066d  
30081                 1             0                           0  01a6602b8  
30082                 0             1                           0  e692d316c  

[30083 rows x 13 columns]

In [6]:
train_data.isna().sum()

StudyInstanceUID              0
ETT - Abnormal                0
ETT - Borderline              0
ETT - Normal                  0
NGT - Abnormal                0
NGT - Borderline              0
NGT - Incompletely Imaged     0
NGT - Normal                  0
CVC - Abnormal                0
CVC - Borderline              0
CVC - Normal                  0
Swan Ganz Catheter Present    0
PatientID                     0
dtype: int64

In [27]:
train_data["PatientID"].describe()

count         30083
unique         3255
top       05029c63a
freq            172
Name: PatientID, dtype: object

## EDA

In [75]:
#Counts the number of patients
train_data["PatientID"].unique().size

3255

Despite there being about 30,000 rows in the table, there are only about 3255 patients. This means that paints will the multiple images of them with varying data points. There are ~9 images per patitent. Lets look further into this.

In [76]:
#Groups the table to log the number of occurrences of each patient
frequencies = pd.DataFrame(train_data.groupby(by="PatientID")["PatientID"].count())
frequencies.rename(columns={"PatientID":"Count"}, inplace=True)

In [77]:
display(frequencies.sort_values(by="Count", ascending=False))
singleImagePatients = frequencies[frequencies["Count"]==1]
mt10ImagePatients = frequencies[frequencies["Count"]>10]
print(f"{singleImagePatients.size} patients have only 1 image")
print(f"{mt10ImagePatients.size} patients have more than 10 images")

Count
PatientID       
05029c63a    172
55073fece    167
26da0d5ad    148
8849382d0    130
34242119f    110
...          ...
c09883b19      1
e46763e02      1
7bcec67b1      1
c0b57c3d9      1
1bc88e955      1

[3255 rows x 1 columns]

397 patients have only 1 image
800 patients have more than 10 images


Knowing this, let's take a look at the patient with the most entries to see if we can root out any stories or patters. How might these catherters change over time in a single patient?

In [91]:
mostImagedPatient = train_data[train_data["PatientID"]=="05029c63a"]
dataToDisplay = mostImagedPatient.sort_values(by= "ETT - Normal")

#Bypasses maximum row limit
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(dataToDisplay)

StudyInstanceUID  ETT - Abnormal  \
2226   1.2.826.0.1.3680043.8.498.74190045333565570017...               0   
12933  1.2.826.0.1.3680043.8.498.61936035308630680796...               0   
28334  1.2.826.0.1.3680043.8.498.83695501915139952540...               0   
24390  1.2.826.0.1.3680043.8.498.42565923266899815850...               0   
10367  1.2.826.0.1.3680043.8.498.89784585798061898325...               0   
14245  1.2.826.0.1.3680043.8.498.93518520610058188214...               0   
10253  1.2.826.0.1.3680043.8.498.81612968658469626927...               0   
27011  1.2.826.0.1.3680043.8.498.14612428778012016836...               0   
17329  1.2.826.0.1.3680043.8.498.51595082650400612709...               0   
4828   1.2.826.0.1.3680043.8.498.74338185865293989379...               0   
1206   1.2.826.0.1.3680043.8.498.68620140014200110095...               0   
16180  1.2.826.0.1.3680043.8.498.20308463414598007724...               0   
29562  1.2.826.0.1.3680043.8.498.87586453518386177872...               0   
8580   1.2.826.0.1.3680043.8.498.95986923920930940143...               0   
21277  1.2.826.0.1.3680043.8.498.56619259631388451199...               0   
21392  1.2.826.0.1.3680043.8.498.68230348425967455500...               0   
21074  1.2.826.0.1.3680043.8.498.81944912067980263193...               0   
21131  1.2.826.0.1.3680043.8.498.77850398282099900559...               0   
438    1.2.826.0.1.3680043.8.498.33289872132944517059...               0   
21545  1.2.826.0.1.3680043.8.498.18520646120291067878...               0   
20923  1.2.826.0.1.3680043.8.498.10795116836379665499...               0   
21620  1.2.826.0.1.3680043.8.498.22708626112974214549...               0   
21736  1.2.826.0.1.3680043.8.498.39791289470110258821...               0   
21758  1.2.826.0.1.3680043.8.498.34011387566126371312...               0   
21869  1.2.826.0.1.3680043.8.498.11349207275419168734...               0   
22120  1.2.826.0.1.3680043.8.498.49556205206559134192...               0   
21537  1.2.826.0.1.3680043.8.498.50449172676528043832...               0   
20514  1.2.826.0.1.3680043.8.498.11263575583722291627...               0   
19803  1.2.826.0.1.3680043.8.498.41227413376741655635...               0   
20295  1.2.826.0.1.3680043.8.498.11299011267454246171...               0   
16344  1.2.826.0.1.3680043.8.498.86724319840387506288...               0   
17216  1.2.826.0.1.3680043.8.498.11173083271185391881...               0   
17253  1.2.826.0.1.3680043.8.498.81300792352709497925...               0   
17448  1.2.826.0.1.3680043.8.498.77921328072083127097...               0   
17672  1.2.826.0.1.3680043.8.498.16685551398126567854...               0   
18380  1.2.826.0.1.3680043.8.498.63763657368460886545...               0   
18384  1.2.826.0.1.3680043.8.498.97598031614637533199...               0   
20393  1.2.826.0.1.3680043.8.498.90530596223907802017...               0   
18447  1.2.826.0.1.3680043.8.498.44716434786763363291...               0   
18719  1.2.826.0.1.3680043.8.498.81891274746132952889...               0   
19065  1.2.826.0.1.3680043.8.498.12897056025467923403...               0   
19199  1.2.826.0.1.3680043.8.498.74749936751008638109...               0   
19317  1.2.826.0.1.3680043.8.498.10463132506595124405...               0   
19645  1.2.826.0.1.3680043.8.498.94144674813450104224...               0   
22214  1.2.826.0.1.3680043.8.498.11941622662325769362...               0   
20176  1.2.826.0.1.3680043.8.498.12572436145733989741...               0   
18654  1.2.826.0.1.3680043.8.498.73217556798203198355...               0   
22471  1.2.826.0.1.3680043.8.498.93131526993837520394...               0   
23259  1.2.826.0.1.3680043.8.498.75659347118535311678...               0   
22638  1.2.826.0.1.3680043.8.498.25754747493697196392...               0   
26541  1.2.826.0.1.3680043.8.498.49233870722947588078...               0   
26576  1.2.826.0.1.3680043.8.498.10813486232359928522...               0   
26852  1.2.8

In [92]:
#This is to analyse a list of all the images of the most-imaged patient
returnAllPatientImages("05029c63a")

['1.2.826.0.1.3680043.8.498.33289872132944517059685300521052847970',
 '1.2.826.0.1.3680043.8.498.58815494070381425972841494897309562888',
 '1.2.826.0.1.3680043.8.498.18423562395321282202151467283211883687',
 '1.2.826.0.1.3680043.8.498.80010847260848855758249955549927520930',
 '1.2.826.0.1.3680043.8.498.68620140014200110095088585785330951729',
 '1.2.826.0.1.3680043.8.498.87375315768078677164756022968175911741',
 '1.2.826.0.1.3680043.8.498.99564257536158506118998101216918000997',
 '1.2.826.0.1.3680043.8.498.43908356782442779725424825386043768718',
 '1.2.826.0.1.3680043.8.498.51991674554454729241003433645598796537',
 '1.2.826.0.1.3680043.8.498.78021892236561877580290458242105043659',
 '1.2.826.0.1.3680043.8.498.86537162817021158811092283918656005072',
 '1.2.826.0.1.3680043.8.498.11525667914743618624335995337909956877',
 '1.2.826.0.1.3680043.8.498.12682264367523788590209252628203092871',
 '1.2.826.0.1.3680043.8.498.74190045333565570017428818377245264093',
 '1.2.826.0.1.3680043.8.498.294983

Let's look at a simple patient, maybe just one in a middle with only 7 images (01a6602b8)

In [94]:
mostImagedPatient = train_data[train_data["PatientID"]=="01a6602b8"]
dataToDisplay = mostImagedPatient

with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(dataToDisplay)

StudyInstanceUID  ETT - Abnormal  \
3547   1.2.826.0.1.3680043.8.498.13368586175752823546...               0   
3834   1.2.826.0.1.3680043.8.498.92443648275106206350...               0   
18199  1.2.826.0.1.3680043.8.498.48575077712160297418...               0   
20679  1.2.826.0.1.3680043.8.498.38434326028461648396...               0   
22063  1.2.826.0.1.3680043.8.498.32265506210757468088...               0   
22775  1.2.826.0.1.3680043.8.498.70674394161629430460...               0   
30081  1.2.826.0.1.3680043.8.498.95092491950130838685...               0   

       ETT - Borderline  ETT - Normal  NGT - Abnormal  NGT - Borderline  \
3547                  0             0               0                 0   
3834                  0             0               0                 0   
18199                 0             0               0                 0   
20679                 0             0               0                 0   
22063                 0             0               0                 0   
22775                 0             0               0                 0   
30081                 0             0               0                 0   

       NGT - Incompletely Imaged  NGT - Normal  CVC - Abnormal  \
3547                           0             0               0   
3834                           0             0               0   
18199                          0             0               0   
20679                          0             0               0   
22063                          0             0               0   
22775                          0             0               0   
30081                          0             0               0   

       CVC - Borderline  CVC - Normal  Swan Ganz Catheter Present  PatientID  
3547                  0             1                           0  01a6602b8  
3834                  0             1                           0  01a6602b8  
18199                 1             0                           0  01a6602b8  
20679                 0             1                           0  01a6602b8  
22063                 0             1                           0  01a6602b8  
22775                 0             1                           0  01a6602b8  
30081                 1             0                           0  01a6602b8

In [95]:
frequencies[frequencies["Count"]==7]

Count
PatientID       
01a6602b8      7
01e3dc988      7
01fa2ad89      7
0261d55c1      7
04323559c      7
...          ...
fc4269a27      7
fcc1c2dfc      7
fce6e7380      7
fed489bad      7
ff826cf51      7

[199 rows x 1 columns]

In [98]:
mostImagedPatient = train_data[train_data["PatientID"]=="ff826cf51"]
dataToDisplay = mostImagedPatient

with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(dataToDisplay)

StudyInstanceUID  ETT - Abnormal  \
9054   1.2.826.0.1.3680043.8.498.85888286754526046422...               0   
16211  1.2.826.0.1.3680043.8.498.50804831665099230965...               0   
17344  1.2.826.0.1.3680043.8.498.29983487174552869646...               0   
17658  1.2.826.0.1.3680043.8.498.73750776498365697385...               0   
18146  1.2.826.0.1.3680043.8.498.12722241245555310535...               0   
20747  1.2.826.0.1.3680043.8.498.65803857161092450324...               0   
25717  1.2.826.0.1.3680043.8.498.12724690186666099182...               0   

       ETT - Borderline  ETT - Normal  NGT - Abnormal  NGT - Borderline  \
9054                  0             0               0                 0   
16211                 0             0               0                 0   
17344                 0             1               0                 0   
17658                 0             1               0                 0   
18146                 0             1               0                 0   
20747                 0             1               0                 0   
25717                 0             0               0                 0   

       NGT - Incompletely Imaged  NGT - Normal  CVC - Abnormal  \
9054                           0             0               0   
16211                          0             0               0   
17344                          1             0               0   
17658                          0             1               0   
18146                          0             1               0   
20747                          1             0               0   
25717                          0             0               0   

       CVC - Borderline  CVC - Normal  Swan Ganz Catheter Present  PatientID  
9054                  0             1                           0  ff826cf51  
16211                 0             1                           0  ff826cf51  
17344                 0             1                           0  ff826cf51  
17658                 0             1                           0  ff826cf51  
18146                 0             1                           0  ff826cf51  
20747                 0             1                           0  ff826cf51  
25717                 1             0                           0  ff826cf51

After taking a peek at some of the patients with only a few images, it can be seen that most images are focused on CVC caherters. Additionally, the CVC catherters seem to bounce back and forth a lot betwtween normal and bordeline and sometimes abnormal. There are two possible conjectures, since there is no record of time between these photos: <br><br>
1] The catherters move about and can change betwwen normal and borerline between images <br>
2] Chaterters start borderline or normal until an event or action which displaces them into another state, in which they stay until disturbed agian. 